src/main/java/com/iexec/common/docker/client/DockerClientInstance.java

```diff
@@ -411,27 +413,41 @@ public DockerRunResponse run(DockerRunRequest dockerRunRequest) {
         if (dockerRunRequest.getMaxExecutionTime() <= 0) {
             // container will run until self-exited or explicitly-stopped
             log.info("Docker container will run in detached mode [name:{}]", containerName);
-            dockerRunResponse.setSuccessful(true);
+            dockerRunResponse.setFinalStatus(DockerRunFinalStatus.SUCCESS);
             return dockerRunResponse;
         }
         Instant timeoutDate = Instant.now()
                 .plusMillis(dockerRunRequest.getMaxExecutionTime());
-        int exitCode = waitContainerUntilExitOrTimeout(containerName, timeoutDate);
-        dockerRunResponse.setContainerExitCode(exitCode);
-        boolean isTimeout = exitCode == -1;
-        boolean isSuccessful = !isTimeout && exitCode == 0L;
-        if (isTimeout && !stopContainer(containerName)) {
-            log.error("Failed to force-stop container after timeout [name:{}]", containerName);
-            return dockerRunResponse;
+        boolean isSuccessful;
+        try {
+            int exitCode = waitContainerUntilExitOrTimeout(containerName, timeoutDate);
+            dockerRunResponse.setContainerExitCode(exitCode);
+
+            isSuccessful = exitCode == 0L;
+            log.info("Finished running docker container [name:{}, isSuccessful:{}]",
+                    containerName, isSuccessful);
+            dockerRunResponse.setFinalStatus(
+                    isSuccessful
+                            ? DockerRunFinalStatus.SUCCESS
+                            : DockerRunFinalStatus.FAILED);
+        } catch (TimeoutException e) {
+            log.error(e.getMessage());
+            dockerRunResponse.setFinalStatus(DockerRunFinalStatus.TIMEOUT);
+            if (!stopContainer(containerName)) {
+                getContainerLogs(containerName).ifPresent(dockerRunResponse::setDockerLogs);
+                log.error("Failed to force-stop container after timeout [name:{}]", containerName);
+                return dockerRunResponse;
+            }
+        } catch (IllegalArgumentException e) {
+            log.error(e.getMessage());
         }
+
         getContainerLogs(containerName).ifPresent(dockerRunResponse::setDockerLogs);
         if (!removeContainer(containerName)) {
             log.error("Failed to remove container after run [name:{}]", containerName);
+            dockerRunResponse.setFinalStatus(DockerRunFinalStatus.FAILED);
```

Comment:
Not sure to get that one :thinking: 
In the case of a timeout, even if the the container wasn't removed, shouldn't the final status be `TIMEOUT`?

Reply:
TBH, not sure about that one, too. Current implementation always returns a failure (`response.isSuccessful() == false`) when removal fails, even though the Docker container successfully exited. If we extend this behavior, even though the Docker container timed out, we should return a failure.
But we can fix this behavior so that it returns a timeout status instead. Let me know what you prefer!

Reply:
I mostly agree with you. I was simply preserving the current implementation. I'd be happy to fix it!

Reply:
By parsing the library, I found that the docker-java StopContainerCmd supports a timeout and can do a forceful kill if the timout is reached.
Shouldn't we lever this to be sure a container is stopped and avoid polluting the user system ?

Reply:
https://github.com/docker-java/docker-java/blob/master/docker-java-api/src/main/java/com/github/dockerjava/api/command/StopContainerCmd.java#L14

Reply:
Thanks @jeremyjams, see https://github.com/iExecBlockchainComputing/iexec-common/pull/336/commits/5116472ce025e83279c2ce2cc5cd91368c1508a6.

Reply:
@jbern0rd > Do you know how this exactly works? If `timeout` is set to 0, wouldn't that always force fill the container?

Reply:
docker-java implementation is a thin wrapper around docker API
https://docs.docker.com/engine/api/v1.41/#operation/ContainerStop
The question es to understand how the docker API works.